### PR TITLE
Move more of `rustc::lint` into `rustc_lint`

### PR DESCRIPTION
**Pull request by Centril**
Based on https://github.com/rust-lang/rust/pull/67806.

Here we try to consolidate more of the linting infra into `rustc::lint`. Some high-level notes:

- We now store an `Lrc<dyn Any + Send + Sync>` as opposed to `Lrc<LintStore>` in the `GlobalCtxt`. This enables us to avoid referring to the type, breaking a cyclic dependency, and so we can move things from `rustc::lint` to `rustc_lint`.

- `in_derive_expansion` is, and needs to, be moved as a method on `Span`.

- We reduce the number of ways on `tcx` to emit a lint so that the developer UX is more streamlined.

- `LintLevelsBuilder` is moved to `rustc_lint::levels`, leaving behind `LintLevelMap/Set` in a purified form due to current constraints (hopefully fixable in the future after https://github.com/rust-lang/rust/pull/68133).

- `struct_lint_level` is moved to `rustc::lint` due to current dependency constraints.

- `rustc::lint::context` is moved to `rustc_lint::context`.

- The visitors in `rustc::lint` are moved to `rustc_lint::passes`.

### CODE REVIEW

**Comment by Centril** [comment]
This is ready for review now.

File: src/librustc_lint/late.rs [hunk 1]
```diff
@@ -31,6 +30,12 @@ use syntax::walk_list;
 use log::debug;
 use std::slice;
 
+/// Extract the `LintStore` from the query context.
+/// This function exists because we've erased `LintStore` as `dyn Any` in the context.
+crate fn unerased_lint_store<'tcx>(tcx: TyCtxt<'tcx>) -> &'tcx LintStore {
```
Comment (Mark-Simulacrum) [inline src/librustc_lint/late.rs]: I guess it probably doesn't matter but it would be nice to return `&'tcx Lrc<LintStore>` here I think, or, if we don't need that then maybe the Lrc can be dropped? (I forget why it exists).

File: src/librustc/lint/levels.rs [hunk 2]
```diff
@@ -159,19 +123,43 @@ pub struct BuilderPush {
...
+        let mut builder = LintLevelsBuilder {
             sess,
-            sets,
+            sets: LintLevelSets::new(),
             cur: 0,
             id_to_set: Default::default(),
             warn_about_weird_lints,
+        };
+        builder.process_command_line(sess, store);
+        assert_eq!(builder.sets.list.len(), 1);
```
Comment (Mark-Simulacrum) [inline src/librustc/lint/levels.rs]: Pre-existing, but maybe stick a FIXME here to drop the Vec<> if we're not using it anywhere?

File: src/librustc/lint.rs [hunk 3]
```diff
@@ -1,30 +1,15 @@
 use std::cmp;
 
 use crate::ich::StableHashingContext;
-use crate::lint::context::CheckLintNameResult;
```
Comment (Mark-Simulacrum) [inline src/librustc/lint.rs]: Seems like git lost the rename here -- maybe we can split the raw move into a separate commit so it keeps track of the rename and we don't lose history?

**Review by Mark-Simulacrum — COMMENTED** [review]
Changes look good.

It'd be helpful to include a description of the new/old crate deps in the PR description I think, but not critical (would help me understand exactly how things have moved at a high level).

File: src/librustc/lint.rs (hunk 3 above)
Comment (Centril) [reply src/librustc/lint.rs]: (We reached the conclusion in private chat that GitHub is being buggy here and that blame is in fact intact.)

File: src/librustc_lint/late.rs (hunk 1 above)
Comment (Centril) [reply src/librustc_lint/late.rs]: (We concluded in private that `&'tcx LintStore` is fine, as it promises less to callers.)

File: src/librustc/lint/levels.rs (hunk 2 above)
Comment (Centril) [reply src/librustc/lint/levels.rs]: Seems like we're using the `Vec<_>` in `fn push`: `self.sets.list.push(LintSet::Node { specs: specs, parent: prev });`.

**Comment by Centril** [comment]
Added a PR description. :)

@bors r=Mark-Simulacrum

File: src/librustc_lint/lib.rs [hunk 4]
```diff
@@ -24,45 +42,47 @@ extern crate rustc_session;
...
 use nonstandard_style::*;
 use redundant_semicolon::*;
-use rustc::lint::internal::*;
 use types::*;
 use unused::*;
 
-/// Useful for other parts of the compiler.
+/// Useful for other parts of the compiler / Clippy.
 pub use builtin::SoftLints;
+pub use context::{EarlyContext, LateContext, LintContext, LintStore};
```
Comment (JohnTitor) [inline src/librustc_lint/lib.rs]: @Centril could we add `CheckLintNameResult` here? Clippy uses it.

Comment (Centril) [reply src/librustc_lint/lib.rs]: Aww... I thought I had made sure all imports Clippy needed was exported, but apparently not. Sure, let's export it. Send me a PR to approve?

Comment (JohnTitor) [reply src/librustc_lint/lib.rs]: Sure!
